src/mlmc/quantity_concept.py

```diff
@@ -121,120 +325,451 @@ def mult_op(x, y):
+        """
+        def mask_gen(x, *args):
+            indices = np.random.choice(x.shape[1], size=size)
+            mask = np.zeros(x.shape[1], bool)
+            mask[indices] = True
+            return mask
+        return self._mask_quantity(size, mask_gen)
+
+    def __getitem__(self, key):
+        """
+        Get items from Quantity, quantity type must support brackets access
+        :param key: str, int, tuple
+        :return: Quantity
+        """
+        new_qtype = self.qtype[key]  # New quantity type
+
+        reshape_shape = None
+        # ArrayType might be accessed directly regardless of qtype start and size
+        if isinstance(self.qtype, ArrayType):
+            slice_key = key
+            reshape_shape = self.qtype._shape
+        # Other accessible quantity types use start and size
+        else:
+            start = new_qtype.start
+            end = new_qtype.start + new_qtype.size()
+            slice_key = slice(start, end)
 
-    def _samples(self, i_chunk):
-        level_chunks = self.storage.sample_pairs(i_chunk)
-        return [chunk[self.begin:self.end, :, :]  for chunk in level_chunks]
+        def getitem_op(y):
+            # Reshape M to original shape to allow access
+            if reshape_shape is not None:
+                y = y.reshape((*reshape_shape, y.shape[-2], y.shape[-1]))
+            y_get_item = y[slice_key]  # indexing
 
+            # Keep dims [M, N, 2]
+            if len(y_get_item.shape) == 2:
+                y_get_item = y_get_item[np.newaxis, :]
+            elif len(y_get_item.shape) > 2:
+                y_get_item = y_get_item.reshape((np.prod(y_get_item.shape[:-2]), y_get_item.shape[-2],
+                                                 y_get_item.shape[-1]))
 
+            return y_get_item
 
+        return Quantity(quantity_type=new_qtype, input_quantities=[self], operation=getitem_op)
 
-class TimeSeries(QBase):
-    """
-    We should assume that all quantities have same structure (check only same size)
-    but can be of any type. Meanigful are only onther types then TimeSeries.
-    """
-    def __init__(self, time_frames : List[Tuple[float, QBase]]):
-        super().__init__(time_frames)
-        self.frame0 = time_frames[0][1]
-        assert( all( (q._size == self.frame0._size for t,q in time_frames) ) )
-        assert( all( (q.__class__ == self.frame0.__class__ for t,q in time_frames) ))
+    def __iter__(self):
+        raise Exception("This class is not iterable")
 
+    def __copy__(self):
+        return Quantity(quantity_type=self.qtype, input_quantities=custom_copy(self._input_quantities),
+                        operation=self._operation)
 
-        self.time_frames = time_frames
-        # list of pairs (time, value)
-        self.block_starts = []
-        # starts of blocks of the time values on the single sample row
-        self.operation = lambda lch : np.concatenate(lch, axis=0)
+    def time_interpolation(self, value):
```

Comment:
I'm not sure this would work for any Quantity type, `self.qtype_times` may be invalid. That should probably be specific to the TimeSeries type.